attestation-agent/kbs_protocol/src/evidence_provider/native.rs

```diff
@@ -3,40 +3,78 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
+use std::collections::HashMap;
+
 use async_trait::async_trait;
-use attester::CompositeAttester;
-use crypto::HashAlgorithm;
-use kbs_types::{Tee, TeePubKey};
+use attester::{detect_attestable_devices, detect_tee_type, BoxedAttester, TeeEvidence};
+use kbs_types::Tee;
 
 use super::EvidenceProvider;
 
 use crate::{Error, Result};
 
-pub struct NativeEvidenceProvider(CompositeAttester);
+pub struct NativeEvidenceProvider {
+    primary_tee: Tee,
+    primary_attester: BoxedAttester,
+    additional_attesters: Vec<(Tee, BoxedAttester)>,
+}
 
 impl NativeEvidenceProvider {
     pub fn new() -> Result<Self> {
-        Ok(Self(
-            CompositeAttester::new().map_err(|e| Error::GetEvidence(e.to_string()))?,
-        ))
+        let primary_tee = detect_tee_type();
+        let primary_attester = BoxedAttester::try_from(primary_tee).map_err(|e| {
+            Error::NativeEvidenceProvider(format!("failed to initialize primary attester: {e}"))
+        })?;
+        let additional_attesters = detect_attestable_devices()
+            .into_iter()
+            .map(|tee| -> Result<_> {
+                let boxed_attester = TryInto::<BoxedAttester>::try_into(tee).map_err(|e| {
+                    Error::NativeEvidenceProvider(format!(
+                        "failed to initialize additional attester: {e}"
+                    ))
+                })?;
+                Ok((tee, boxed_attester))
+            })
+            .collect::<Result<Vec<(Tee, BoxedAttester)>>>()?;
+        Ok(Self {
+            primary_tee,
+            primary_attester,
+            additional_attesters,
+        })
     }
 }
 
 #[async_trait]
 impl EvidenceProvider for NativeEvidenceProvider {
-    async fn get_evidence(
-        &self,
-        tee_pubkey: TeePubKey,
-        nonce: String,
-        hash_algorithm: HashAlgorithm,
-    ) -> Result<String> {
-        self.0
-            .composite_evidence(tee_pubkey, nonce, hash_algorithm)
+    async fn primary_evidence(&self, runtime_data: Vec<u8>) -> Result<TeeEvidence> {
+        self.primary_attester
+            .get_evidence(runtime_data)
             .await
             .map_err(|e| Error::GetEvidence(e.to_string()))
     }
 
+    async fn get_additional_evidences(&self, runtime_data: Vec<u8>) -> Result<String> {
```

Comment:
nit: sadly the plural of evidence is evidence, so we should probably change all these.

Reply:
Oops. Let me fix this